File: 2_csharp/Aoc.cs
using System;
using System.Collections.Generic;

namespace aoc
{
    internal class Aoc
    {
        public static int PartOne(int[] input) => Run(input, 12, 2)[0];

        public static int PartTwo(int[] input)
        {
            for (var noun = 0; noun <= 99; noun++)
            for (var verb = 0; verb <= 99; verb++)
            {
                var inputCopy = new int[input.Length];
                Array.Copy(input, inputCopy, input.Length);

                if (Run(inputCopy, noun, verb)[0] == 19690720)
                    return 100 * noun + verb;
            }

            throw new ArgumentException("Couldn't find what we're looking for in input");
        }

        private static int[] Run(int[] input, int noun, int verb)
        {
            input[1] = noun;
            input[2] = verb;

            return Run(input);
        }

        public static int[] Run(int[] input)
        {
            var ip = 0;

            while (true)
                switch (input[ip])
                {
                    case 1:
                        ip = Operator(input, ip, (x, y) => x + y);
                        break;
                    case 2:
                        ip = Operator(input, ip, (x, y) => x * y);
                        break;
                    case 99:
                        return input;
                }
        }

        private static int Operator(IList<int> input, int ip, Func<int, int, int> op)
        {
            input[input[ip + 3]] = op(input[input[ip + 1]], input[input[ip + 2]]);
            return ip + 4;
        }
    }
}

File: 2_elixir/lib/aoc.ex
defmodule Aoc do
  def part_one(input),
    do: run(input, 12, 2)

  def part_two(input),
    do: part_two(input, 99, 99)

  def part_two(input, noun, -1),
    do: part_two(input, noun - 1, 99)

  def part_two(input, noun, verb) do
    case run(input, noun, verb) do
      19_690_720 -> 100 * noun + verb
      _ -> part_two(input, noun, verb - 1)
    end
  end

  def to_indexed_map(input),
    do:
      0..length(input)
      |> Enum.zip(input)
      |> Map.new()

  def run(input, noun, verb),
    do:
      %{(input |> to_indexed_map()) | 1 => noun, 2 => verb}
      |> run(0)
      |> Map.get(0)

  def run(input),
    do:
      input
      |> to_indexed_map()
      |> run(0)
      |> Map.values()

  def run(input, ip) do
    case input[ip] do
      1 -> run(input |> operator(ip, &+/2), ip + 4)
      2 -> run(input |> operator(ip, &*/2), ip + 4)
      99 -> input
    end
  end

  def operator(input, ip, op),
    do:
      input
      |> Map.merge(%{
        input[ip + 3] => op.(input[input[ip + 1]], input[input[ip + 2]])
      })
end
